kf_task_fhir_etl/etl/ingest.py

```diff
@@ -0,0 +1,525 @@
+            )
+            descendants["studies"] = study
+            if investigator is not None:
+                descendants["investigators"] = investigator
+
+            # Cache a study in memory
+            snapshot[kf_study_id] = descendants
+            found += 1
+
+        assert expected == found, f"Found {found} study(ies) but expected {expected}"
+
+        return snapshot
+
+    def extract(self):
+        """Extracts records.
+
+        :return: A dictionary mapping an endpoint to records
+        :rtype: dict
+        """
+        snapshot = self._create_snapshot(self.kf_study_ids)
+        mapped_df_dict = defaultdict()
+
+        for kf_study_id, descendants in snapshot.items():
+            logging.info(f"  ⏳ Extracting {kf_study_id}")
+
+            # Loop over descendants
+            for endpoint, records in descendants.items():
+                df = None
+                if endpoint in {"investigators", "studies"}:
+                    df = records
+                else:
+                    df = pd.DataFrame.from_dict(records, orient="index")
+                df = df.drop(columns=["uuid", "created_at", "modified_at"])
+                mapped_df_dict.setdefault(kf_study_id, {})[endpoint] = df
+                logging.info(f"    📁 {endpoint} {df.shape}")
+
+            logging.info(f"  ✅ Extracted {kf_study_id}")
+
+        return mapped_df_dict
+
+    def transform(self, mapped_df_dict):
+        """Transforms records.
+
+        :param mapped_df_dict: An output from the above exract stage
+        :type mapped_df_dict: dict
+        :return: A dictionary of outer-merged data frames
+        :rtype: dict
+        """
+        merged_df_dict = defaultdict()
+
+        for kf_study_id, study_mapped_df_dict in mapped_df_dict.items():
+            logging.info(f"  ⏳ Transforming {kf_study_id}")
+            merged_df_dict.setdefault(kf_study_id, {})
+            study_merged_df, study_all_targets = None, set()
+
+            # studies
+            studies = study_mapped_df_dict["studies"]
+            studies = studies.rename(
+                columns={
+                    "investigator_id": CONCEPT.INVESTIGATOR.TARGET_SERVICE_ID,
+                    "attribution": CONCEPT.STUDY.ATTRIBUTION,
+                    "data_access_authority": CONCEPT.STUDY.AUTHORITY,
+                    "domain": "STUDY|DOMAIN",
+                    "external_id": CONCEPT.STUDY.ID,
+                    "kf_id": CONCEPT.STUDY.TARGET_SERVICE_ID,
+                    "name": CONCEPT.STUDY.NAME,
+                    "program": "STUDY|PROGRAM",
+                    "release_status": CONCEPT.STUDY.RELEASE_STATUS,
+                    "short_code": "STUDY|SHORT_CODE",
+                    "short_name": CONCEPT.STUDY.SHORT_NAME,
+                    "version": CONCEPT.STUDY.VERSION,
+                    "visible": CONCEPT.STUDY.VISIBLE,
+                }
+            )
+            study_all_targets.add(ResearchStudy)
```

Comment:
Just wondering why all of these `study_all_targets.add/update` aren't done in one place at the beginning or end of the loop? Might be easier to read

Reply:
This is to control: 

- The ingest order (there are dependencies between endpoints, for instance, `/Condition` can't get it before `/Patient`); and
- Adding conditional endpoints to the load queue (See below for more details).

Reply:
I think you could still do the add/update in the right order if they were all at the beginning of the loop, but I think what you have is fine